### PR TITLE
Swarm mode with python env

### PR DESCRIPTION
**Pull request by nsosio**
Follow up on [344](https://github.com/premAI-io/prem-app/pull/344). This PR replace pyinstaller and binaries for running petals swarm by:
- installing miniconda at `~/.config/prem` with bash script `src-tauri/petals/create_env.sh`
- creating a miniconda environment at `~/.config/prem/envs/prem-env`
- installing the requirements `src-tauri/petals/requirements.txt`
- and running petals

### CODE REVIEW

**Review by casperdcl — COMMENTED** [review]
I'd propose moving some of this logic to `prem-services` & `prem-registry`, i.e. having a swarm-as-a-service.

1. `prem-app` installs miniconda at `~/.config/prem`
2. `prem-services` defines a `petals_swarm.sh` script which does
   ```sh
   ${PREM_PYTHON:-python} install -r requirements.txt
   ${PREM_PYTHON:-python} main.py
   ```
3. `prem-registry` points to the above service
   ```josn
   "serviceType": "python",
   "serveCommand": "./petals_swarm.sh"
   ```
4. `prem-app` points to the above registry
5. `prem-app` will do one of the following (depending on whether we want isolation):
   1. `conda env create -n {manifest.id} && conda activate {manifest.id} && PREM_PYTHON=~/.config/prem/envs/{manifest.id}/bin/python {manifest.serveCommand}`
   2. `PREM_PYTHON=~/.config/prem/bin/python {manifest.serveCommand}`

**Comment by tiero** [comment]
Consider adding a bash script to clean the environment, ie. uninstall miniconda and any other file/artifact generated by Prem installation.

**Comment by filopedraz** [comment]
The swarm mode is VERY particular. How can Prem App knows which is the service for Swarm mode? We need extra logic in the frontend in order to parse the services. Moreover, the swarm mode is triggered from Settings page. Let's keep it simple for now.

File: src/modules/settings/components/SwarmMode.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  }, []);
+
+  useEffect(() => {
+    swarmSupported().then(setIsSwarmSupported);
+  }, []);
+
+  const onStart = async (e: React.FormEvent) => {
+    try {
+      e.preventDefault();
+      await runSwarmMode(numBlocks, model, "premAI");
```
Comment (casperdcl) [inline src/modules/settings/components/SwarmMode.tsx:40]: probably want to wrap this line with a UI spinner

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -99,6 +106,156 @@ fn is_docker_running() -> bool {
...
+    let _ = Command::new("sh")
+        .args([format!("{petals_path}/create_env.sh")])
+        .envs(env)
+        .output()
+        .expect("🙈 Failed to create env");
+    format!("{app_data_dir}/envs/prem-env/bin/python3").to_string()
+}
+
+
+#[tauri::command]
```
Comment (casperdcl) [inline src-tauri/src/main.rs:206]: as @swarnimarun recommended, looks like this is enough :)

```suggestion
#[tauri::command(async)]
```

File: src/modules/settings/components/SwarmMode.tsx (hunk 1 above)
Comment (nsosio) [reply src/modules/settings/components/SwarmMode.tsx:40]: any suggestions on how to wrap it with a spinner?

Comment (casperdcl) [reply src/modules/settings/components/SwarmMode.tsx:40]: looks like they're [used in a few places](https://github.com/search?q=repo%3ApremAI-io%2Fprem-app%20Spinner&type=code)

Comment (nsosio) [reply src/modules/settings/components/SwarmMode.tsx:50]: done :)

**Comment by filopedraz** [comment]
### Missing

- [ ] Tooltips for the fields. Users don't know Petals.
- [ ] Modal: `Congratulations, you are now part of the Swarm. Check it out here: https://explorer.prem.ninja`

**Comment by Janaka-Steph** [comment]
Rather a toast than a modal no?

**Comment by filopedraz** [comment]
The toast disappears. People don't have time to copy and paste the link.

File: src-tauri/src/main.rs [hunk 3]
```diff
@@ -99,6 +106,174 @@ fn is_docker_running() -> bool {
...
+        ])
+        .spawn()
+        .expect("🙈 Failed to run swarm");
+}
+
+
+fn get_swarm_processes() -> String {
+    // Check if create_env.sh is running
+    let output = Command::new("/usr/bin/pgrep")
+        .args(&["-f", "create_env.sh|conda.exe"])
```
[the remark this answers is not in the document]
Comment (nsosio) [reply src-tauri/src/main.rs:229]: By inspecting the running processes there was `conda.exe` running but I can relax it to be `conda`.

File: src-tauri/src/main.rs [hunk 4]
```diff
@@ -99,6 +106,174 @@ fn is_docker_running() -> bool {
     return false;
 }
 
+#[tauri::command]
+fn is_swarm_supported() -> bool {
+    match env::consts::OS {
+        "macos" => true,
```
Comment (casperdcl) [inline src-tauri/src/main.rs:112]: I guess we allow Linux too based on the current `create_env.sh` & `requirements.txt`

File: src-tauri/tauri.conf.json [hunk 5]
```diff
@@ -22,11 +22,17 @@
         "exit": true
       },
       "shell": {
-        "all": true
+        "all": true,
+        "execute": true,
+        "sidecar": true,
+        "open": true
```
[the remark this answers is not in the document]
Comment (nsosio) [reply src-tauri/tauri.conf.json]: done :)

**Comment by filopedraz** [comment]
Great @nsosio. If you still have some time, I suggest you start with the rebase :)

**Comment by filopedraz** [comment]
### Some comments

- [ ] There is a small overlay (shadow) around the "form", we should remove it
- [ ] The tooltip is good, but we should maybe have also a small label on top of it. It's not clear enough right now.
- [ ] While the spinner runs, we should have a small label on the left saying `Configuring the Swarm Environment...`.
- [ ] The install state is not persisted right now. If too complicated, no problem, we can skip that for now.
- [ ] The link in the modal should be clickable and text should be more informative. 

We need to increase verbosity in general because it's a complex feature and people will be completely mis-lead if they try it out.

PS: we don't need to handle all the details in this PR. We can merge and then have separate issues and small fixes. I can also help out with the copy.

File: src-tauri/petals/create_env.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+prem_appdir="${PREM_APPDIR:-.}"
+requirements="$(dirname "$0")/requirements.txt"
+conda_prefix="$prem_appdir/miniconda"
+
+if test ! -x "$conda_prefix/bin/mamba"; then
+  echo "Conda not found"
+  os="$(uname)"
+  case "$os" in
+    Darwin|Linux)
```
Comment (casperdcl) [inline src-tauri/petals/create_env.sh:10]: are you sure mac outputs `Darwin` when you run `uname` with no args?

https://github.com/conda-forge/miniforge#unix-like-platforms-mac-os--linux seems to imply it should be `MacOSX`...

File: .github/workflows/on-push-server.yaml [hunk 7]
```diff
@@ -1,42 +1,32 @@
...
-          --file Dockerfile
-          --tag ghcr.io/premai-io/prem-app:latest
-          --platform linux/arm64,linux/amd64 .
+    - uses: actions/checkout@v4
+    - name: metadata
+      id: metadata
+      run: |
+        git fetch --all --tags
+        TAG=$(git describe --tags)
+        echo "tag=${TAG/-*}" >> $GITHUB_OUTPUT
```
Comment (casperdcl) [inline .github/workflows/on-push-server.yaml:16]: @tiero I added this tag logic - does it make sense?

**Review by casperdcl — COMMENTED** [review]
I've just mega-rebased onto current `main`. I'm sure there are a few bugs. Things to do @nsosio:

- [ ] update `src-tauri/Cargo.lock`
- [ ] update `package-lock.json`
- [ ] check for bugs (e.g. missing `use` declarations)

Original backup branch (pre-mega-rebase): https://github.com/premAI-io/prem-app/tree/nsosio/feat/swarm-mode-bash

**Comment by tiero** [comment]
This crate fixes the env as in pulling the .bashrc of the user, we could avoid using absolute paths if necessary https://github.com/tauri-apps/fix-path-env-rs

**Comment by casperdcl** [comment]
I noticed @nsosio added then removed it for some reason

**Comment by nsosio** [comment]
@casperdcl I've added it because wget was giving problems but we are not using it anymore and I've removed it.

**Comment by nsosio** [comment]
- [x]  Removed a small overlay (shadow) around the "form"
- [x]  Added labels for swarm parameters
- [x]  Added label while creating environment
- [x]  Persistent state for swarm
- [x]  Link to explorer clickable

File: src-tauri/src/swarm.rs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            &model,
+        ])
+        .spawn()
+        .expect("🙈 Failed to run swarm");
+}
+
+pub fn get_swarm_processes() -> String {
+    // Check if create_env.sh is running
+    let output = Command::new("/usr/bin/pgrep")
+        .args(&["-f", "create_env.sh|conda.exe"])
```
Comment (casperdcl) [inline src-tauri/src/swarm.rs:133]: what about this?

```suggestion
        .args(&["-f", "create_env.sh|(mamba|conda).*create.*prem_app"])
```

**Comment by tiero** [comment]
When `Configuring the Swarm Environment...` shall we add "Please do not close the app?"

**Comment by tiero** [comment]
FYI with cargo tauri dev

```
```

File: src/shared/helpers/utils.ts [hunk 9]
```diff
@@ -10,6 +11,12 @@ import type { ControlProps, CSSObjectWithLabel } from "react-select";
 
 export const SERVICE_CHECK_REFETCH_INTERVAL = 10000;
 
+export const petalsModels = async (): Promise<string[]> => {
+  const models = await invoke("get_petals_models");
+  console.log(models);
```
Comment (tiero) [inline src/shared/helpers/utils.ts:16]:
```suggestion
```

**Comment by tiero** [comment]
I personally still stuck on `Configuring the Swarm Environment...` even cleaning up in `rm -rf ~/.config/prem`

**Comment by filopedraz** [comment]
Did you remove `/target` directory too?

**Comment by casperdcl** [comment]
btw @nsosio we should [split `create_env.sh` into two files](https://github.com/premAI-io/prem-registry/pull/88#issuecomment-1782613906) - this first file really just creates an env (`env create -n prem_app`), the second file installs & starts the swarm (`pip install && main.py`)

**Comment by tiero** [comment]
tACK https://github.com/premAI-io/prem-app/pull/393/commits/935f990698f521e87b30034e37763a6fedcb941b

Ready to merge!

**Comment by nsosio** [comment]
@tiero why on swarm start we link to https://network.premai.io and not to https://explorer.prem.ninja in the modal?

**Comment by tiero** [comment]
> @tiero why on swarm start we link to https://network.premai.io and not to https://explorer.prem.ninja in the modal?

This is the actual production endpoint

**Pull request merged**